File: src/client/src/vcs/history.ts
import { Dev } from "../common/dev";

/**
 */
export interface Update {
	title: string;
	explainer: string;
	type: Update.Type;
	peopleToTell: Update.PeopleToTell;
	script: {
		code?: string;
		data?: string;
	}; // TODO
}
export namespace Update {
	export type Id = number;
	export enum Type {
		Addition, // opt-in optimizations do not go here.
		Correction_Minor,
		Correction_Urgent,
		ExplainerEdit,
		Rephrase, // ie. refactor
		Optimization,
	};
	export interface PeopleToTell {
		contributors: boolean;
		codeUsers: boolean;
		appUsers: boolean;
	};
}

export interface UpdateHistory {
	updates: Update[];
}


export interface UpdateDraft {
	title: string; // must be unique on the server.
	base: Update.Id;
	explainer: string;
	changes: UpdateDraft.Change[];

	/**
	 * Fancy magic. Ex. remove a change if a later change renders it useless.
	 *
	 * Or perhaps this is best left to the developer?
	 */
	normalize(): void;
}
export namespace UpdateDraft {

	export interface Change {
		kind: Change.Kind;
		body: any;
		explainer: string;
		author: Dev;
	}
	export namespace Change {
		export enum Kind {
			Add,
			Edit,
			Remove,
		}
	}
}

File: packages/gemel/lib/src/vcs/history.dart
typedef UpdateId = int;

///
class Update {
  String title;
  String explainer;
  UpdateType type;
  UpdatePeopleToTell peopleToTell;
  Object scriptCode; // TODO
  Object scriptData; // TODO

  Update({
    required this.title,
    required this.explainer,
    required this.type,
    required this.peopleToTell,
    required this.scriptCode,
    required this.scriptData,
  });
}

///
enum UpdateType {
  addition, // opt-in optimizations do not go here.
  correctionMinor,
  correctionUrgent,
  explainerEdit,
  rephrase, // ie. refactor
  optimization,
}

///
class UpdatePeopleToTell {
  bool contributors;
  bool codeUsers;
  bool appUsers;

  UpdatePeopleToTell({
    required this.contributors,
    required this.codeUsers,
    required this.appUsers,
  });
}

///
class UpdateHistory {
  List<Update> updates;

  UpdateHistory({
    required this.updates,
  });
}

///
class UpdateDraft {
  String title; // must be unique on the server.
  UpdateId base;
  String explainer;
  List<UpdateDraftChange> changes;

  UpdateDraft({
    required this.title,
    required this.base,
    required this.explainer,
    required this.changes,
  });

  /// Fancy magic. Ex. remove a change if a later change renders it useless.
  ///
  /// Or perhaps this is best left to the developer?
  void normalize() {}
}

class UpdateDraftChange {
  UpdateDraftChangeKind kind;
  Object body;
  String explainer;
  Dev author;

  UpdateDraftChange({
    required this.kind,
    required this.body,
    required this.explainer,
    required this.author,
  });
}

enum UpdateDraftChangeKind {
  add,
  edit,
  remove,
}
